### PR TITLE
BE-031 Filtering, sorting and pagination for discounts (Elizaveta Dubogryz)

### PR DESCRIPTION
**Pull request by el-dub**

### CODE REVIEW

**Review by rudik-byte — DISMISSED** [review]

**Review by mblack666 — DISMISSED** [review]
Where QDiscount in Entity?

File: src/main/java/com/exadel/discount/service/impl/DiscountServiceImpl.java [hunk 1]
```diff
@@ -144,4 +150,22 @@ private Vendor findVendor(UUID vendorId) {
                 .findByIdAndArchived(vendorId, false)
                 .orElseThrow(() -> new NotFoundException(String.format("Vendor with ID %s doesn't exist", vendorId)));
     }
-}
\ No newline at end of file
+
+    private Predicate getPredicate(DiscountFilter filter) {
```
Comment (SergeyKoval) [inline src/main/java/com/exadel/discount/service/impl/DiscountServiceImpl.java]:
```suggestion
    private Predicate preparePredicateForFindingAll(DiscountFilter filter) {
```

File: src/main/java/com/exadel/discount/util/QPredicates.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package com.exadel.discount.util;
```
Comment (SergeyKoval) [inline src/main/java/com/exadel/discount/util/QPredicates.java]:
```suggestion
package com.exadel.discount.repository.query;
```

File: src/main/java/com/exadel/discount/util/QPredicates.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import com.querydsl.core.types.Predicate;
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Function;
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public class QPredicates {
```
Comment (SergeyKoval) [inline src/main/java/com/exadel/discount/util/QPredicates.java]:
```suggestion
public class QueryPredicateBuilder {
```

File: src/main/java/com/exadel/discount/util/QPredicates.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+    public Predicate buildAnd() {
+        return ExpressionUtils.allOf(predicates);
+    }
+
+    public Predicate buildOr() {
+        return ExpressionUtils.anyOf(predicates);
+    }
+
+    public static QPredicates builder() {
```
Comment (SergeyKoval) [inline src/main/java/com/exadel/discount/util/QPredicates.java]:
```suggestion
    public static QPredicates init() {
```

File: src/main/java/com/exadel/discount/util/QPredicates.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Function;
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public class QPredicates {
+    private List<Predicate> predicates = new ArrayList<>();
+
+    public <T> QPredicates add(T object, Function<T, Predicate> function) {
```
Comment (SergeyKoval) [inline src/main/java/com/exadel/discount/util/QPredicates.java]:
```suggestion
    public <T> QPredicates append(T object, Function<T, Predicate> function) {
```

**Review by SergeyKoval — CHANGES_REQUESTED** [review]

File: src/main/java/com/exadel/discount/service/impl/DiscountServiceImpl.java (hunk 1 above)
Comment (el-dub) [reply src/main/java/com/exadel/discount/service/impl/DiscountServiceImpl.java]: +

File: src/main/java/com/exadel/discount/util/QPredicates.java (hunk 2 above)
Comment (el-dub) [reply src/main/java/com/exadel/discount/util/QPredicates.java]: +

File: src/main/java/com/exadel/discount/util/QPredicates.java (hunk 3 above)
Comment (el-dub) [reply src/main/java/com/exadel/discount/util/QPredicates.java]: +

File: src/main/java/com/exadel/discount/util/QPredicates.java (hunk 4 above)
Comment (el-dub) [reply src/main/java/com/exadel/discount/util/QPredicates.java]: +

File: src/main/java/com/exadel/discount/util/QPredicates.java (hunk 5 above)
Comment (el-dub) [reply src/main/java/com/exadel/discount/util/QPredicates.java]: +

**Review by rudik-byte — APPROVED** [review]

**Review by SergeyKoval — APPROVED** [review]

**Pull request merged**